### ISSUE TITLE
Einsum should have an `out=` parameter

### ISSUE THREAD

**Issue by thomasahle** [pytorch/pytorch#75701]
### 🚀 The feature, motivation and pitch

When using `einsum` in performance sensitive code (and not caring about gradients), it is not good that it allocates a new tensor for the result. Then `numpy` version has an `out=` parameter, similar to `bmm` and many other pytorch functions, which allow specifying a piece of memory to store the result. It would be good to have this in pytorch as well.

### Alternatives

Alternatively users can avoid einsum entirely, but there is not always a pytorch function with the same functionality.

**Comment by thomasahle** [comment]
Is there a general issue tracking adding `out=` parameters to more functions?
It would also be useful to have the option in `torch.diagonal`. It already exists in `torch.diag`, but only `.diagonal` works on batched matrices.

**Comment by thomasahle** [comment]
I ran into one more: `numpy.sum` has an `out=`, but `torch.sum` doesn't.

**Comment by gchanan** [comment]
> It would also be useful to have the option in torch.diagonal. It already exists in torch.diag, but only .diagonal works on batched matrices.

This doesn't help with the memory issue right, because it is returning a view?

**Comment by jeshels** [comment]
`resize` and `interpolate` functions can benefit from this as well.

**Comment by thomasahle** [comment]
> > It would also be useful to have the option in torch.diagonal. It already exists in torch.diag, but only .diagonal works on batched matrices.
> This doesn't help with the memory issue right, because it is returning a view?
Ah, I didn't actually know "partial views" were a thing. Searching for "pytorch partial view" this notion seems to only be used for the torch.diagonal function? I suppose the `out=` variant would copy the data instead.

**Comment by thomasahle** [comment]
I just realized `torch.bmm(A, B)` has an `out=` parameter, but `A.bmm(B)` doesn't. That can probably be added to the list.

**Comment by Lezcano** [comment]
@thomasahle  Just free functions have out parameters so this is expected.
